### PR TITLE
#11, #16, #18, #20: Update project implementation and wiki documentation.

### PR DESCRIPTION
**Pull request by dustincatap**
Add/update implementation of #11, #16, #18, #20

### CODE REVIEW

File: docs/Code-Generation.md [hunk 1]
```diff
@@ -0,0 +1,104 @@
+We'll use [`build_runner`](https://pub.dev/packages/build_runner) to generate the code for:
+
+- Dependency injection using [`injectable`](https://pub.dev/packages/injectable)
+- JSON serialization/deserialization using [`json_serializable`](https://pub.dev/packages/json_serializable)
+- Resource management using [`flutter_gen`](https://pub.dev/packages/flutter_gen)
+- Mapper classes using [`auto_mappr`](https://pub.dev/packages/auto_mappr)
+- Localization files using [`intl_utils`](https://pub.dev/packages/intl_utils)
+- Mocks in tests using [`mockito`](https://pub.dev/packages/mockito)
+- Database schema using [`isar`](https://pub.dev/packages/isar)
+- API clients using [`retrofit``](https://pub.dev/packages/retrofit)
```
Comment (ThomasKaegi-Erni) [inline docs/Code-Generation.md:10]: One uptick too many.

File: docs/Code-Generation.md [hunk 2]
```diff
@@ -0,0 +1,104 @@
...
+    <String, dynamic>{
+      'Email': instance.email,
+      'Password': instance.password,
+    };
+
+```
+
+:exclamation: **<span style="color: red">IMPORTANT</span>**
+
+- These generated files are **not** commited to git. Code eneration should be part of your CI workflow.
```
Comment (ThomasKaegi-Erni) [inline docs/Code-Generation.md:104]: I find that to be an excellent decision :-)

File: docs/Home.md [hunk 3]
```diff
@@ -0,0 +1,37 @@
...
+## Getting Started
+
+- [Install Flutter](https://docs.flutter.dev/get-started/install)
+- [Introduction to Dart Language](https://dart.dev/guides/language/language-tour)
+- [Flutter Documentation](https://docs.flutter.dev/)
+
+## Prerequisites
+
+**Flutter**
+- v3.13.5+ on stable channel
```
Comment (ThomasKaegi-Erni) [inline docs/Home.md:20]: Just a thought here: Refering to a specific version will be outdated quite soon, especially for a starterkit application that is not constantly being maintained...
I suggest to ask the reader to install the latest stable version. Then if that causes trouble to open a new issue here. That way the starterkit will more likely be kept up to date, as we'll be getting feedback from users.

File: docs/Project-Setup.md [hunk 4]
```diff
@@ -0,0 +1,43 @@
...
+Clone the repo.
+
+```sh
+git clone https://github.com/ERNI-Academy/starterkit-mobile-application-flutter.git
+```
+
+Untrack the files under the folder `.secrets` by adding it in your `.gitignore` file.
+
+```sh
+.secrets/**
```
Comment (ThomasKaegi-Erni) [inline docs/Project-Setup.md:12]: I'd suggest to add that to this repo's gitignore file. As the (default) secrets are already checked in, they won't get lost.
And if we maintainers see that the defaults need an update, we can always force add the changes.

File: docs/Project-Setup.md
Comment (ThomasKaegi-Erni) [inline docs/Project-Setup.md:1]: Nice work, by the way! So far really good setup documentation :-)

File: docs/Unit-Testing.md [hunk 5]
```diff
@@ -0,0 +1,80 @@
+## Naming Convention
+
+| Items        | Naming Pattern                                                     | Target        | Example                                                |
```
Comment (ThomasKaegi-Erni) [inline docs/Unit-Testing.md:3]: Nice, to see the test naming convention formally defined.

File: docs/core/Environments.md [hunk 6]
```diff
@@ -0,0 +1,88 @@
...
+    "iOSDevelopmentTeam": "XXXXXXXXXX",
+    "iOSDevelopmentProfile": "Your Development Profile Name",
+    "iOSDistributionProfile": "Your Distribution Profile Name",
+    "iOSExportMethod": "ad-hoc"
+}
+```
+
+:exclamation: **<span style="color: red">IMPORTANT</span>**
+
+- The folder `.secrets` is committed to git by default (for testing purposes). You should **delete** the `.secrets` folder and  **uncomment** last part of the **.gitignore** file to remove it from git. Once you have done this, you can start adding your secrets files locally.
```
Comment (ThomasKaegi-Erni) [inline docs/core/Environments.md:44]: As suggested earlier, I reckon it'd be better to ignore the secrets folder by default.

File: docs/core/Environments.md [hunk 7]
```diff
@@ -0,0 +1,88 @@
...
+- For iOS, you need to add the necessary provisioning profiles and signing certificates to your local machine. You also need to supply the correct `iOSDevelopmentTeam` value in the secrets file and in the **`ios/Flutter/DartDefine-Defaults.xcconfig`** file.
+- The secrets file are also used in Fastlane. See `fastlane/Fastfile` for more details.
```
Comment (ThomasKaegi-Erni) [inline docs/core/Environments.md:46]: Consider making this an actual link.

File: docs/core/Environments.md [hunk 8]
```diff
@@ -0,0 +1,88 @@
...
+
+### Accessing Dart Defines in the Platforms
+
+The Android and iOS native projects are already configured to update their current configurations based on the environment file we will use. For example, the provisioning profile used for running the app on an iPhone device are automatically selected (given that the profile is available in your local machine).
+
+You can access other Dart defines in the native projects by referencing the key directly. 
+
+An example of this can be found in `ios/Runner/Info.plist`:
```
Comment (ThomasKaegi-Erni) [inline docs/core/Environments.md:54]: Consider making this an actual link.

File: docs/data/Database.md [hunk 9]
```diff
@@ -0,0 +1,80 @@
...
+
+## Creating a Local Data Source
+
+`LocalDataSource` is an interface that defines the operations that can be done in the database. We use this interface to hide the implementation details of the database to the consumers of the data source.
+
+```dart
+abstract interface class LocalDataSource<T extends DataObject> {
+  Future<T?> get(int id);
+
+  Future<Iterable<T>> getAll();
```
Comment (ThomasKaegi-Erni) [inline docs/data/Database.md:44]: From a design perspective I'd encourage never to include a `getAll()` method in sample code, a base class or an interface definition of a DB access layer. Bad things will happen as soon as we target a production DB that will likely hold "a lot" of data.

`getAll` methods on DB access layers are IMHO only ok for tables that are small and will remain small by design. (eg. "category" tables etc.)

For "regular" tables `get` operations should awlays be done with some kind of filter, selector or predicate to limmit the data retrieved to what we're actually interested in.

File: docs/data/Database.md [hunk 10]
```diff
@@ -0,0 +1,80 @@
...
+
+  Future<void> addOrUpdate(T object);
+
+  Future<void> addOrUpdateAll(Iterable<T> objects);
+
+  Future<void> delete(T object);
+
+  Future<void> deleteAll();
```
Comment (ThomasKaegi-Erni) [inline docs/data/Database.md:52]: Similar to the above `deleteAll` methods are equally dangerous if not even more dangerous. Consider removing this  from the sample code.

File: docs/domain/models/Entities.md
Comment (ThomasKaegi-Erni) [inline docs/domain/models/Entities.md:1]: Compared to the other docs this file is quite empty. Are you planning to add more context?

File: docs/presentation/State-Management.md [hunk 11]
```diff
@@ -0,0 +1,53 @@
+## ValueNotifiers, ValueListenable and ValueListenableBuilders
+
+Currently, there are [multiple state management packages](https://docs.flutter.dev/development/data-and-backend/state-mgmt/options).
+
+Although not mentioned in Flutter's recommended approaches, [`ValueListenable`](https://api.flutter.dev/flutter/foundation/ValueListenable-class.html)/[`ValueNotifier`](https://api.flutter.dev/flutter/foundation/ValueNotifier-class.html)combined with [`ValueListenableBuilder`](https://api.flutter.dev/flutter/widgets/ValueListenableBuilder-class.html) proved to be much simpler and is readily available out-of-the-box. This approach is very simple and allows us to update only the needed part of the UI.
+
+Given the following view model:
+
+```dart
+class HomeViewModel extends ViewModel {
```
Comment (ThomasKaegi-Erni) [inline docs/presentation/State-Management.md:10]: I love this pattern :-)

File: docs/presentation/View-Models.md [hunk 12]
```diff
@@ -0,0 +1,129 @@
...
+}
+```
+
+- The `ViewModel` class inherits from `ChangeNotifier`.
+- A new instance is returned always whenever we resolve a view model.
+- To know how to extract parameters passed during navigation, read about it [here](navigation).
+
+<div style="">
+
+:bulb: **<span style="color: green">TIP</span>**
```
Comment (ThomasKaegi-Erni) [inline docs/presentation/View-Models.md:24]: I like your use of `TIP`, `AVOID`, etc patterns. As it provides a quick entry into the most relevant parts of the docs.

File: starterkit_app/.vscode/launch.json [hunk 13]
```diff
@@ -29,7 +29,7 @@
             },
             "args": [
                 "--dart-define-from-file",
-                ".secrets/${input:appEnvironment}.json",
```
Comment (ThomasKaegi-Erni) [inline starterkit_app/.vscode/launch.json:32]: Was this change intentional? It seems to contradict your docs.

File: starterkit_app/lib/core/presentation/views/view_life_cycle_handler.dart [hunk 14]
```diff
@@ -14,7 +14,7 @@ import 'package:starterkit_app/core/reflection.dart';
 import 'package:starterkit_app/core/service_locator.dart';
 
 abstract final class ViewLifeCycleHandler {
-  static NavigationObserver get _navigationObserver => ServiceLocator.instance<NavigationObserver>();
+  static final NavigationObserver _navigationObserver = ServiceLocator.instance<NavigationObserver>();
```
Comment (ThomasKaegi-Erni) [inline starterkit_app/lib/core/presentation/views/view_life_cycle_handler.dart:17]: In the PR diff, there's a Codecov mention that the changed line is not covered by tests.
Perfect :-)
That's one of the configurations you changed in the starterkit, right?

File: starterkit_app/test/assets/post.json
Comment (ThomasKaegi-Erni) [inline starterkit_app/test/assets/post.json:1]: FYI: This file amongst others is missing an empty last line. It'd be great we add that last line...

**Review by ThomasKaegi-Erni — COMMENTED** [review]
Great work. IMHO the starterkit has improved in expressiveness and quality. I have some general comments that you may or may not choose to integrate in this PR.

The only comment I have that I think we "should" discuss is the 'getAll' method in the DB access layer...

File: docs/domain/models/Entities.md
Comment (dustincatap) [reply docs/domain/models/Entities.md:1]: I'll add more contents in this one. There are still missing documentations for some parts of the project, I'll be adding them sooner.

File: starterkit_app/.vscode/launch.json (hunk 13 above)
Comment (dustincatap) [reply starterkit_app/.vscode/launch.json:32]: Yes this is intentional. I originally used the actual secrets file when running tests. I changed it and used a test asset file `test/assets/test_secrets.json` instead.

File: starterkit_app/lib/core/presentation/views/view_life_cycle_handler.dart (hunk 14 above)
Comment (dustincatap) [reply starterkit_app/lib/core/presentation/views/view_life_cycle_handler.dart:17]: Yes. I recently added Codecov.

File: starterkit_app/test/database/test_isar_database_factory.dart [hunk 15]
```diff
@@ -0,0 +1,32 @@
+import 'dart:io';
+
+import 'package:isar/isar.dart';
+import 'package:path/path.dart';
+import 'package:starterkit_app/core/data/database/isar_database_factory.dart';
+
+// Download the latest libraries from https://github.com/isar/isar/releases
```
[the remark this answers is not in the document]
Comment (dustincatap) [reply starterkit_app/test/database/test_isar_database_factory.dart:7]: I'll check on this. On the top of my head, I'm thinking of using a custom script that will run every time we run/debug tests via VSCode.

It will:

1. Check if the libraries are existing.
2. Check if their version matches the same version of `Isar` in `pubspec.yaml`.
3. Download or replace the files if they are missing or out of date, for the current platform on which the tests will be running (Windows, macOS, or Linux).

I can integrate this as well as part of the CI for tests. and we can completely remove the library files from git.

File: docs/Project-Setup.md
Comment (dustincatap) [reply docs/Project-Setup.md:1]: As an additional, I created a simple bash script named `create.sh` that will clone and rename the project and all of its files according to the user's desired configurations.

File: starterkit_app/test/database/test_isar_database_factory.dart (hunk 15 above)
[the remark this answers is not in the document]
Comment (ThomasKaegi-Erni) [reply starterkit_app/test/database/test_isar_database_factory.dart:7]: That would be a great way of automating this. If you think this is out of scope of your masterpriece, feel free to just open an issue to resolve this later.

File: docs/data/Database.md (hunk 9 above)
Comment (dustincatap) [reply docs/data/Database.md]: I have updated the implementation to include a selector in the form of `offset` and `limit`.

I also updated the `Database.md` documentation by adding `Extending the Local Data Source` section, which explains the usage of other queries that get objects using a specific filter for that table.

**Review by ThomasKaegi-Erni — APPROVED** [review]
This is IMHO a significant step toward a high quality starterkit :-)
Nicely done :-)
